src/nnp_acsf.F

```diff
@@ -0,0 +1,1336 @@
+            ! dr1/dx1 = -dr1/dx2
+            tmp = pref*symtmp*ftot
+            tmp1 = pref*angular*(ftot*dsymdr1 + dfcutdr1*symtmp)
+            tmp2 = pref*angular*(ftot*dsymdr2 + dfcutdr2*symtmp)
+            tmp3 = pref*angular*(ftot*dsymdr3 + dfcutdr3*symtmp)
+            DO i = 1, 3
+               force(i, 1, sf) = tmp*dangulardx1(i) + tmp1*dr1dx(i) + tmp2*dr2dx(i)
+               force(i, 2, sf) = tmp*dangulardx2(i) - tmp1*dr1dx(i) + tmp3*dr3dx(i)
+               force(i, 3, sf) = tmp*dangulardx3(i) - tmp2*dr2dx(i) - tmp3*dr3dx(i)
+            END DO
+         END IF
+
+         ! Don't forget prefactor: 2**(1-ang%zeta)
+         pref = prefzeta
+         ! combine angular and exponential part with cutoff function
+         sym(sf) = pref*angular*symtmp*ftot
+      END DO
+
+   END SUBROUTINE nnp_calc_ang
+
+! ***************************************************************************
+!> \brief Sort element array according to atomic number
+!> \param ele ...
+!> \param nuc_ele ...
+!> \date   2020-10-10
+!> \author Christoph Schran (christoph.schran@rub.de)
+! **************************************************************************************************
+   SUBROUTINE nnp_sort_ele(ele, nuc_ele)
+      CHARACTER(len=2), ALLOCATABLE, DIMENSION(:)        :: ele
+      INTEGER, ALLOCATABLE, DIMENSION(:)                 :: nuc_ele
+
+      CHARACTER(len=*), PARAMETER :: routineN = 'nnp_sort_ele', routineP = moduleN//':'//routineN
+
+      CHARACTER(len=2)                                   :: tmp_ele
+      INTEGER                                            :: i, j, loc, minimum, tmp_nuc_ele
+
+      ! Determine atomic number
+      DO i = 1, SIZE(ele)
+         CALL get_ptable_info(ele(i), number=nuc_ele(i))
+      END DO
+
+      ! Sort both arrays
+      DO i = 1, SIZE(ele) - 1
+         minimum = nuc_ele(i)
+         loc = i
+         DO j = i + 1, SIZE(ele)
+            IF (nuc_ele(j) .LT. minimum) THEN
+               loc = j
+               minimum = nuc_ele(j)
+            END IF
+         END DO
+         tmp_nuc_ele = nuc_ele(i)
+         nuc_ele(i) = nuc_ele(loc)
+         nuc_ele(loc) = tmp_nuc_ele
+
+         tmp_ele = ele(i)
+         ele(i) = ele(loc)
+         ele(loc) = tmp_ele
+
+      END DO
+
+   END SUBROUTINE nnp_sort_ele
+
+! ***************************************************************************
+!> \brief Sort symmetry functions according to different criteria
+!> \param nnp ...
+!> \date   2020-10-10
+!> \author Christoph Schran (christoph.schran@rub.de)
+! **************************************************************************************************
+   SUBROUTINE nnp_sort_acsf(nnp)
+      TYPE(nnp_type), POINTER                            :: nnp
+
+      CHARACTER(len=*), PARAMETER :: routineN = 'nnp_sort_acsf', routineP = moduleN//':'//routineN
+
+      INTEGER                                            :: i, j, k, loc
+
+      ! First sort is according to symmetry function type
```

Comment:
We actually have a generic [sort implementation](https://github.com/cp2k/cp2k/blob/master/src/common/array_sort.fypp).

All you have to do is [include](https://github.com/cp2k/cp2k/blob/master/src/task_list_methods.F#L78) it and define a [less_than routine](https://github.com/cp2k/cp2k/blob/master/src/task_list_methods.F#L2882-L2909).

Reply:
This is the remaining TODO from the comments above. I'll push a first commit that mainly addresses the coding conventions and will look into the sorting afterwards.

Reply:
It seems this PR is essentially read to be merged. Should I wait for the sorting or do you want to open a separate PR for that?

Reply:
I think a second PR would actually be nicer. That gives me a little bit more time to focus on other things and I can bundle the sorting with additional improvements that will come up eventually (as mentioned above).

Thanks a lot for your feedback on the PR!